### PR TITLE
Ports - Steph

### PR DESCRIPTION
**Pull request by smarchante1**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The Model keeps track of application data and handles business logic. It communicates to the applications controller.
Describe in your own words what the Controller is doing in Rails | The Controller handles web server requests and pulls information the model. It contains actions the define things for it to do.
Describe in your own words what the View is doing in Rails | Presents information to the user using ERB.
Describe an edge-case controller test you wrote | An edge case test I wrote was to test redirects when handling actions on a task id that does not exist.
What is the purpose of using strong params? (i.e. the params method in the controller) |  Strong params protect sensitive model attributes from being updated by users at will.
How are Rails migrations related to Rails models? |  Migrations change the the structure of models and provide a schema for the database.
Describe one area of Rails that are still unclear on | I still need practice with managing migrations; specifically what to do when you need to make an edit to your database. I have an idea, but it's a fuzzy one.

### CODE REVIEW

File: app/views/tasks/_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+  <%= f.label :name %>
+  <%= f.text_field :name %>
+  <%= f.label :completion_date %>
+  <%= f.text_field :completion_date %>
+  <%= f.label :description %>
+  <%= f.text_field :description %>
+
+
+  <%= f.submit "edit task", class: "task-form__submit-btn" %>
```
Comment (CheezItMan) [inline app/views/tasks/_form.html.erb:13]: I suggest passing a parameter into the partial, like we did in Ada-books, so the new and edit forms have different text on the submit button.

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -79,50 +74,105 @@
...
+    #     name: "Home",
+    #     description: "Walk the dog",
+    #     completion_date: "April, 11th 2019",
+    #   )
+
+    #   patch complete_task_path(task.id)
+
+    #   toggle = Task.find_by(id: task.id)
+    
+    #   expect(toggle.completion_date).must_equal "April, 11th 2019"
```
Comment (CheezItMan) [inline test/controllers/tasks_controller_test.rb:173]: This is going to need to do something like:
```ruby
      expect(DateTime.parse(toggle.completion_date).to_i).must_be_close_to  DateTime.now.to_i, 5
```

This is because your TasksController is saving a DateTime into the string field.  

Note you should also test the `completed` field

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  def index
+    @tasks = Task.all
+  end
+
+  def show
+    task_id = params[:id].to_i
+    @task = Task.find_by(id: task_id)
+
+    if @task.nil?
+      redirect_to action: 'index', status: 302
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:12]: The 302 status code means `found`, which isn't the case here.

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  end
+
+  def update
+    task = Task.find_by(id: params[:id])
+
+    if task.nil?
+      redirect_to tasks_path
+    else  
+      task.update(task_params)
+      flash[:success] = "Successful Update!"
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:48]: 👍

File: app/controllers/tasks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    task = Task.find_by(id: params[:id])
+
+    if task.completed
+      task.update(
+        completed: !task.completed
+      )
+    else
+      task.update(
+        completed: !task.completed,
+        completion_date: DateTime.now
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:64]: Look at this and then the test you commented out.

File: app/controllers/tasks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    task.save
+    redirect_to tasks_path
+  end
+
+
+  def destroy
+    task = Task.find_by(id: params[:id])
+    unless task
+      head :not_found
+      return
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:77]: Generally I'd suggest just having an `else` for the code where task was found, and not having the extra return in here.

File: test/controllers/tasks_controller_test.rb [hunk 7]
```diff
@@ -79,50 +74,105 @@
...
+
+    # end
+
```
Comment (CheezItMan) [inline test/controllers/tasks_controller_test.rb:176]: I would also write a test to verify that you can unmark a task as complete by doing the patch request twice.

**Comment by CheezItMan** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | Check
Answered comprehension questions | Check, hopefully this week will help you with the migration confusion.
Successfully handles: Index, Show | Check
Index & Show tests pass | Check
Successfully handles: New, Create |  Check
New & Create tests pass | Check
Successfully handles: Edit, Update | Check
Tests for Edit & Update test valid & invalid task ids |  Check
Successfully handles: Destroy, Task Complete |  Check
Tests for Destroy & Task Complete include tests for valid and invalid task ids |  Check
Routes follow RESTful conventions |  Check
Uses named routes (like `_path`) |  Check
**Overall** | Well done, you hit all the learning goals.  I gave some feedback, mostly for the tests around marking the task complete.  Let me know if you have questions.
